pkg/keycloak/client.go

```diff
@@ -38,637 +32,368 @@ type Client struct {
 	res, err := c.requester.Do(req)
 	if err != nil {
 		logrus.Errorf("error on request %+v", err)
-		return errors.Wrap(err, "error performing create client request")
+		return errors.Wrapf(err, "error performing UPDATE %s request", resourceName)
 	}
 
-	if res.StatusCode != 204 {
-		return errors.New("failed to update client: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
+	if res.StatusCode < 200 || res.StatusCode > 299 {
+		logrus.Errorf("failed to UPDATE %s %v", resourceName, res.Status)
+		return errors.New(fmt.Sprintf("failed to UPDATE %s: (%d) %s", resourceName, res.StatusCode, res.Status))
 	}
 
+	logrus.Debugf("response:", res)
 	return nil
 }
 
-func (c *Client) ListClients(realmName string) ([]*v1alpha1.KeycloakClient, error) {
-	req, err := http.NewRequest(
-		"GET",
-		fmt.Sprintf("%s/auth/admin/realms/%s/clients", c.URL, realmName),
-		nil,
-	)
-	if err != nil {
-		return nil, err
-	}
-
-	req.Header.Add("Authorization", "Bearer "+c.token)
-	res, err := c.requester.Do(req)
-	if err != nil {
-		logrus.Errorf("error on request %+v", err)
-		return nil, errors.Wrap(err, "error performing clients list request")
-	}
-
-	if res.StatusCode < 200 || res.StatusCode > 299 {
-		return nil, errors.New("failed to list clients: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
-	}
-
-	body, err := ioutil.ReadAll(res.Body)
-	if err != nil {
-		logrus.Errorf("error reading response %+v", err)
-		return nil, errors.Wrap(err, "error reading clients list response")
-	}
-
-	clients := []*v1alpha1.KeycloakClient{}
-	if err := json.Unmarshal(body, &clients); err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal clients list")
-	}
-
-	return clients, nil
+func (c *Client) UpdateRealm(kcRealm, objRealm *v1alpha1.KeycloakRealm) error {
+	err := c.update(objRealm, fmt.Sprintf("realms/%s", kcRealm.ID), "realm")
+	return err
 }
 
-func (c *Client) CreateUser(user *v1alpha1.KeycloakUser, realmName string) error {
-	body, err := json.Marshal(user)
-	if err != nil {
-		return nil
-	}
-
-	req, err := http.NewRequest(
-		"POST",
-		fmt.Sprintf("%s/auth/admin/realms/%s/users", c.URL, realmName),
-		bytes.NewBuffer(body),
-	)
-	if err != nil {
-		return err
-	}
-
-	req.Header.Add("Content-Type", "application/json")
-	req.Header.Add("Authorization", "Bearer "+c.token)
-	res, err := c.requester.Do(req)
-	if err != nil {
-		logrus.Errorf("error on request %+v", err)
-		return errors.Wrap(err, "error performing create user request")
-	}
+func (c *Client) UpdateClient(kcClient, objClient *v1alpha1.KeycloakClient, realmName string) error {
+	err := c.update(objClient, fmt.Sprintf("realms/%s/clients/%s", realmName, kcClient.ID), "client")
+	return err
+}
 
-	if res.StatusCode < 201 || res.StatusCode > 409 {
-		return errors.New("failed to create user: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
-	}
+func (c *Client) UpdateUser(kcUser, objUser *v1alpha1.KeycloakUser, realmName string) error {
```

Comment:
Can you change all `objUser`s and `objClient`s to `specUser` etc for consistency?

Reply:
Why is kcUser here? Thought we removed that.